controller/deployer/strategies/main.go

```diff
@@ -39,60 +141,87 @@ func jobEventsEqual(expected, actual jobEvents) bool {
 	return true
 }
 
-type jobEvents map[string]map[string]int
-
-func waitForJobEvents(events chan *ct.JobEvent, deployEvents chan<- ct.DeploymentEvent, releaseID string, expected jobEvents, log log15.Logger) error {
+func (d *Deploy) waitForJobEvents(releaseID string, expected jobEvents, log log15.Logger) error {
 	actual := make(jobEvents)
-outer:
+
+	handleEvent := func(typ, state string) {
+		if _, ok := actual[typ]; !ok {
+			actual[typ] = make(map[string]int)
+		}
+		actual[typ][state] += 1
+		d.deployEvents <- ct.DeploymentEvent{
+			ReleaseID: releaseID,
+			JobState:  state,
+			JobType:   typ,
+		}
+	}
+
 	for {
-	inner:
 		select {
-		case event, ok := <-events:
+		case event := <-d.serviceEvents:
+			if id, ok := event.Instance.Meta["FLYNN_APP_ID"]; !ok || id != d.AppID {
+				continue
+			}
+			if id, ok := event.Instance.Meta["FLYNN_RELEASE_ID"]; !ok || id != releaseID {
+				continue
+			}
+			typ, ok := event.Instance.Meta["FLYNN_PROCESS_TYPE"]
+			if !ok {
+				continue
+			}
+			if _, ok := d.useJobEvents[typ]; ok {
+				continue
+			}
+			log.Info("got service event", "type", typ, "state", event.Kind)
+			switch event.Kind {
+			case discoverd.EventKindUp:
+				handleEvent(typ, "up")
+			case discoverd.EventKindDown:
+				handleEvent(typ, "down")
```

Comment:
I think we need to always use job events to track the "down" event as an instance could go down in discoverd and then come back up.